py/desispec/scripts/tile_redshifts.py

```diff
@@ -190,28 +179,49 @@ def generate_tile_redshift_scripts(group, nights=None, tileid=None, expids=None,
     # - NOTE: depending upon options, this might re-read all the exptables again
     # - NOTE: this may not scale well several years into the survey
     if group == 'cumulative':
+        if nights is not None:
+            lastnight = int(np.max(nights))
+        elif exptable is not None:
+            lastnight = int(np.max(exptable['NIGHT']))
+        else:
+            lastnight = None
         log.info(f'{len(tileids)} tiles; searching for exposures on prior nights')
-        global _allexp
-        if _allexp is None:
-            log.info(f'Reading all exposure_tables from all nights')
-            _allexp = read_minimal_exptables_columns()
-        keep = np.in1d(_allexp['TILEID'], tileids)
-        newexptable = _allexp[keep]
+        log.info(f'Reading all exposure_tables from all nights')
+        newexptable = read_minimal_science_exptab_cols(tileids=tileids)
+        newexptable = newexptable[['EXPID', 'NIGHT', 'TILEID']]
+
         if exptable is not None:
             expids = exptable['EXPID']
             missing_exps = np.in1d(expids, newexptable['EXPID'], invert=True)
             if np.any(missing_exps):
-                latest_exptable = read_minimal_exptables_columns(nights=np.unique(exptable['NIGHT'][missing_exps]))
-                keep = np.in1d(latest_exptable['EXPID'], expids[missing_exps])
-                latest_exptable = latest_exptable[keep]
-                newexptable = vstack([newexptable, latest_exptable])
+                log.warning(f'Identified {np.sum(missing_exps)} missing exposures '
+                            + f'in the exposure cache. Resetting the cache to acquire'
+                            + f' them from all nights')
+                ## reset_cache will remove cache but it won't be repopulated
+                ## unless we request all nights. So let's request all nights
+                ## then subselect to the nights we want
+                latest_exptable = read_minimal_science_exptab_cols(tileids=tileids,
+                                                                   reset_cache=True)
```

Comment:
This is another place that smells like external code has to know details about how the cache works.  Is this only an optimization, or is the reset_cache=True vs. False really necessary?

Reply:
This is a rephrasing of what was already in the code. I believe the reason for it was that this wants to bring in all exposures including the ones we are submitting at the time of submission here. So the concern was that the cache wouldn't have the last exposure. I _think_ the new cache should actually be robust to this such that it isn't necessary, but rather than add unintended and hard to identify bugs I left the logic as it was (albeit with a different piece of code). We can revisit this after K1 and do some more deliberate tests to see if it is necessary.